### PR TITLE
Magazine Tabbed Carousel with autoscroll #604

### LINKED ISSUE #604: FR: Magazine Tabbed Carousel (#1 of 4)
From vg-macktrucks-com-rd created by [Gonzalo-CG](https://github.com/Gonzalo-CG): Netcentric/vg-macktrucks-com-rd#144  

### Description
A new block, similar to the [Tabbed Carousel Block](https://github.com/Netcentric/vg-macktrucks-com-rd/issues/25) will be developed in order to achieve the design named in Figma as [Group 700](https://www.figma.com/file/oNwe2jHIICPMIe551EY62l/Mack-Reskin---Full-Project?type=design&node-id=2222-50879&mode=dev).

> The option to add a video to tabs and the auto-scrolling of tabs will be handled separately on issue Netcentric/vg-macktrucks-com-rd#208 

The design is the following:
Screenshot of design

### Acceptance Criteria
- [ ] The Dark variant of the Tabbed Carousel Block is available for authors to use on any page.
- [ ] The Dark variant of the Tabbed Carousel Block has the same design for Desktop & mobile as shown in Figma.
- [ ] As an author, I can configure the link that's displayed on every tab. (The Driven By Design text overlapping the media is a link). This text is manually added by the author
- [ ] The block is one tabbed carousel component - Not split into two blocks 
- [ ] Each item links to a magazine article
- [ ] The category and title of an article are pulled from the metadata
- [ ] If the title of an article is too long, it gets truncated with an ellipsis (to be defined when this happens)
- [ ] The first step of this block is to contain images and not video, which will get taken up in #607 

### Open Questions
- ~~Is this a tabbed carousel, or 2 separate components (a hero and some buttons below)?~~ **--> This is just one "tabbed-carousel" component.**
- ~~Is the "Driven by Design" text a link?~~ --> **Yes**
- ~~Does every tab of the carousel have the "subscribe" button overlayed on the top-right corner?~~ **--> Yes**
- ~~Does this block auto-scroll?~~ --> **Yes**

### PR DESCRIPTION
**Pull request by santi-homps**
Fix #604

Test URLs:
- Before: https://develop--vg-macktrucks-com--hlxsites.hlx.page/drafts/shomps/v2-magazine-tabbed-carousel
- After: https://604-mgz-tabbed-carousel--vg-macktrucks-com--hlxsites.hlx.page/drafts/shomps/v2-magazine-tabbed-carousel

### CODE REVIEW

**Comment by jonatan-lledo-netcentric** [comment]
In the demo page, if there are less than 4 tabs, the scroll animation breaks. Also, if there is more than 1 block, the scrolling effect doesn't work always for all the blocks there

File: blocks/v2-magazine-tabbed-carousel/v2-magazine-tabbed-carousel.js [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+const blockName = 'v2-magazine-tabbed-carousel';
+let autoScroll = true;
+let scrollIntervalID;
+const autoScrollInterval = 2000;
+let isFirstLoad = true;
+
+const updateActiveItem = (elements, entry) => {
+  elements.forEach((el, index) => {
+    if (el === entry.target && entry.intersectionRatio >= 0.75) {
+      const carouselItems = el.parentElement;
```
Comment (amglebihan) [inline blocks/v2-magazine-tabbed-carousel/v2-magazine-tabbed-carousel.js:13]: Consider splitting this function into smaller ones. We could extract the code and create a function that updates the active state from here...

File: blocks/v2-magazine-tabbed-carousel/v2-magazine-tabbed-carousel.js [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+      const navigation = el.parentElement.nextElementSibling;
+
+      [carouselItems, navigation].forEach((c) => c.querySelectorAll('.active').forEach((i) => i.classList.remove('active')));
+      carouselItems.children[index].classList.add('active');
+      navigation.children[index].classList.add('active');
+
+      // Center navigation item
```
Comment (amglebihan) [inline blocks/v2-magazine-tabbed-carousel/v2-magazine-tabbed-carousel.js:20]: ... and create another that handles centring the navigation item from here.

File: blocks/v2-magazine-tabbed-carousel/v2-magazine-tabbed-carousel.js [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+  const allArticles = await getJsonFromUrl(url);
+  const articleArray = Object.values(allArticles.data);
+
+  const carouselContainer = createElement('div', { classes: `${blockName}__container` });
+  const carouselItems = createElement('ul', { classes: `${blockName}__items` });
+  carouselContainer.append(carouselItems);
+
+  const tabNavigation = createElement('ul', { classes: `${blockName}__navigation` });
+
+  function buildTabNavigation(title, category, index) {
```
Comment (amglebihan) [inline blocks/v2-magazine-tabbed-carousel/v2-magazine-tabbed-carousel.js:47]: Use an arrow function to be consistent with the rest of the code.

File: blocks/v2-magazine-tabbed-carousel/v2-magazine-tabbed-carousel.js [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        navigation.scrollTo({
+          left: scrollPosition,
+          behavior: 'smooth',
+        });
+      }
+    }
+  });
+};
+
+export default async function decorate(block) {
```
Comment (amglebihan) [inline blocks/v2-magazine-tabbed-carousel/v2-magazine-tabbed-carousel.js:36]: A suggestion, and same as above, I would try to make this code more modular. The decorate function should initialize the carousel, and delegate specific tasks to smaller functions. It contains a lot of code at the moment. That would make the code easier to read and to maintain. As we can use Copilot on the project, you could use it to give you hints on how to break down the code.

**Comment by amglebihan** [comment]
- When the auto scrolling mode is on and you click on an item, it seems that it switches too fast to the next tab.
- Also, the styling of the active item changes in a weird manner sometimes.
- Should we handle other background colors? I guess that we could make use of CSS variables in more places if it should be the case.

File: blocks/v2-magazine-tabbed-carousel/v2-magazine-tabbed-carousel.js [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    };
+    if (enabled) {
+      scrollIntervalID = setInterval(autoScrollFunction, autoScrollInterval);
+    } else {
+      clearInterval(scrollIntervalID);
+      scrollIntervalID = null;
+    }
+  };
+
+  const autoScrollSwitch = document.createRange().createContextualFragment(`
```
Comment (amglebihan) [inline blocks/v2-magazine-tabbed-carousel/v2-magazine-tabbed-carousel.js:120]: It seems that the buttons are not accessible when tabbing. I would also add some styling for the focus-visible state of the buttons (same as the back to top inside the footer for example).
```
outline: 2px solid var(--border-focus);
outline-offset: 2px;
```

File: blocks/v2-magazine-tabbed-carousel/v2-magazine-tabbed-carousel.js [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+const allArticles = await getJsonFromUrl(url);
+const articleArray = Object.values(allArticles.data);
+
+const updateActiveItem = (elements, entry) => {
+  elements.forEach((el, index) => {
+    if (el === entry.target && entry.intersectionRatio >= 0.75) {
+      const carouselItems = el.parentElement;
+      const navigation = el.parentElement.nextElementSibling;
+
+      [carouselItems, navigation].forEach((c) => c.querySelectorAll('.active').forEach((i) => i.classList.remove('active')));
```
Comment (alexiscoelho) [inline blocks/v2-magazine-tabbed-carousel/v2-magazine-tabbed-carousel.js:24]: Not 100% sure about the practice you guys have been following on this project but it seems that you are using BEM then this `active` class should be a modifier. Maybe something like:
`v2-magazine-tabbed-carousel__navigation-item--active`

Comment (alexiscoelho) [reply blocks/v2-magazine-tabbed-carousel/v2-magazine-tabbed-carousel.js:24]: Also, store the class name in a variable and reuse it across the code might be better

**Review by alexiscoelho — CHANGES_REQUESTED** [review]

**Review by alexiscoelho — APPROVED** [review]

**Review by amglebihan — APPROVED** [review]

**Pull request merged**